man/tpm2_verifydigestsignature.1.md
% tpm2_verifydigestsignature(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_verifydigestsignature**(1) - Verify a signature over a message digest with a 
TPM key.

# SYNOPSIS

**tpm2_verifydigestsignature** [*OPTIONS*]

# DESCRIPTION

**tpm2_verifydigestsignature**(1) - This command verifies a signature over a 
precomputed digest using a TPM verification key.

The key referenced by **keyHandle** must use a signing scheme that supports 
signing a digest, such as **TPM_ALG_ECDSA**



**NOTE**: Unrestricted **ML-DSA** keys can only be used with **TPM2_SignDigest()**
if **allowExternalMu** is TRUE.

# OPTIONS

  * **-c**, **\--key-context**=_OBJECT_:

    Reference to the public key used for the signature verification.

  * **-g**, **\--context**=_FILE_:
  
    Optionnal additional verification context.
    
  * **-d**, **\--digest**=_FILE_:
  
    The digest whose signature is to be verified.
    
  * **-s**, **\--signature**=_FILE_:
  
    Input file containing the signature to verify.
    
  * **-t**, **\--validation**=_FILE_:
  
    Output file path for the returned validation ticket.
    
    On success, the ticket tag will be **TPM_ST_DIGEST_VERIFIED**.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

## Setup
The first step is to create the primary object with mldsa.

```bash
tpm2 createprimary -C o -G mldsa87 -g sha512 -c signkey.ctx -a 'fixedtpm|fixedparent|sensitivedataorigin|userwithauth|sign'
```

Step 2 is to sign.

```bash
tpm2 signdigest -c signkey.ctx -d digest.bin -o signature.bin
```

Step 3 is to verify the signature.

```bash
tpm2 verifydigestsignature -c signkey.ctx -d digest.bin -s signature.bin -t validation.bin
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_verifydigestsignature.c
/* SPDX-License-Identifier: BSD-3-Clause */
/***********************************************************************;
* Copyright (c) 2026, STMicroelectronics
*
* All rights reserved.
***********************************************************************/

#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_convert.h"
#include "tpm2_tool.h"
#include "object.h"
#include "pcr.h"
#include "tpm2_alg_util.h"
#include "tpm2_auth_util.h"
#include "tpm2_hierarchy.h"
#include "tpm2_options.h"

typedef struct tpm_verifydigestsignature_ctx tpm_verifydigestsignature_ctx;
struct tpm_verifydigestsignature_ctx {
	
    struct{
		const char *ctx_path;
		tpm2_loaded_object object;
	}key;
	
	const char *context_path;
	TPM2B_SIGNATURE_CTX context;
	const char *digest_path;
	TPM2B_DIGEST digest;
	const char *signature_path;
	TPMT_SIGNATURE signature;
	
	const char *validation_path;
	TPMT_TK_VERIFIED *validation;
};

static tpm_verifydigestsignature_ctx ctx = {
	.context ={
		.size = 0,
	},
	.digest = {
		.size = 0,
	},
	.signature = {0} ,
	.validation = NULL,
};

static bool load_signature_context(const char *path, TPM2B_SIGNATURE_CTX *context){
	
	if (!path){
		context->size = 0;
		return true;
	}
	
	unsigned long size = 0;
	bool result = files_get_file_size_path(path, &size);
	if (!result){
		LOG_ERR("Unable to get signature context file: %s", path);
		return false;
	}
	
	if (size > sizeof(context->buffer)){
		LOG_ERR("Signature context too large: got %lu max %lu", size, sizeof(context->buffer));
		return false;
	}
	
	context->size = size;
	return files_load_bytes_from_path(path, context->buffer, &context->size);
	
}

static bool load_digest(const char *path, TPM2B_DIGEST *digest){
	
	if (!path){
		LOG_ERR("Expected digest input file");
		return false;
	}
	
	unsigned long size = 0;
	bool result = files_get_file_size_path(path, &size);
	if (!result){
		LOG_ERR("Unable to get digest file size: %s", path);
		return false;
	}
	
	if (size > sizeof(digest->buffer)){
		LOG_ERR("Digest too large: got %lu max %zu", size, sizeof(digest->buffer));
		return false;
	}
	
	digest->size = size;
	return files_load_bytes_from_path(path, digest->buffer, &digest->size);
	
}

static bool load_signature_file(const char *path, TPMT_SIGNATURE *signature){
	
	FILE *f = fopen(path, "rb");
	if (!f){
		LOG_ERR("Cannot open file");
		return false;
	}
	
	UINT8 buffer[8192];	
	size_t size_file = fread(buffer, 1, sizeof(buffer), f);	
	fclose(f);
	
	size_t offset = 0;
	TSS2_RC rval = Tss2_MU_TPMT_SIGNATURE_Unmarshal(buffer, size_file, &offset, signature);
	if (rval != TSS2_RC_SUCCESS){
		return false;
	}
	return true;
	
}

static tool_rc check_options(ESYS_CONTEXT *ectx){
	
	UNUSED(ectx);
	
	if (!ctx.key.ctx_path) {
		LOG_ERR("Expected options -c for key object");
		return tool_rc_option_error;
	}
	
	if (!ctx.digest_path) {
		LOG_ERR("Expected options -d for input digest");
		return tool_rc_option_error;
	}
	
	if (!ctx.signature_path) {
		LOG_ERR("Expected options -s for signature output");
		return tool_rc_option_error;
	}
	
	if (!ctx.validation_path) {
		LOG_ERR("Expected options -t for signature output");
		return tool_rc_option_error;
	}
	
	return tool_rc_success;
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.key.ctx_path,
		NULL, &ctx.key.object, false,
		TPM2_HANDLE_ALL_W_NV);
	
	if (rc != tool_rc_success){
		return rc;
	}
	
	bool result = load_signature_context(ctx.context_path, &ctx.context);
	if(!result){
		return tool_rc_general_error;
	}
	
	result = load_digest(ctx.digest_path, &ctx.digest);
	if(!result){
		return tool_rc_general_error;
	}
	
	result = load_signature_file(ctx.signature_path, &ctx.signature);
	if(!result){
		return tool_rc_general_error;
	}
	
	return tool_rc_success;
}
    
static tool_rc verify_digest_signature(ESYS_CONTEXT *ectx){
		
	TSS2_RC rval = Esys_VerifyDigestSignature(ectx, ctx.key.object.tr_handle,
		ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
		&ctx.context, &ctx.digest, &ctx.signature, &ctx.validation);
		
	if (rval != TPM2_RC_SUCCESS){
		LOG_PERR(Esys_VerifyDigestSignature, rval);
		return tool_rc_from_tpm(rval);
	}
	
	return tool_rc_success;
}

static tool_rc process_output(ESYS_CONTEXT *ectx){
	
	UNUSED(ectx);
	
	if(ctx.validation_path) {
		if (!files_save_ticket(ctx.validation, ctx.validation_path)) {
			return tool_rc_general_error;
		}
	}
	
	return tool_rc_success;
}

static bool on_option(char key, char *value) {

    switch (key) {
    case 'c':
        ctx.key.ctx_path = value;
        break;
    case 'g':
		ctx.context_path = value;
		break;
	case 'd':
		ctx.digest_path = value;
		break;
    case 's':
        ctx.signature_path = value;
        break;
    case 't':
        ctx.validation_path = value;
        break;
    default:
		return false;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static const struct option topts[] = {
        { "key-context",       required_argument, NULL, 'c' },
        { "context",           required_argument, NULL, 'g' },
        { "digest",            required_argument, NULL, 'd' },
        { "signature",         required_argument, NULL, 's' },
        { "validation",        required_argument, NULL, 't' }
    };


    *opts = tpm2_options_new("c:g:d:s:t:", ARRAY_LEN(topts), topts, on_option,
            NULL, 0);

    return *opts != NULL;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
    
    UNUSED(flags);
    
    tool_rc rc = check_options(ectx);
    if (rc != tool_rc_success){
		return rc;
	}

    rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }
    
    rc = verify_digest_signature(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    return process_output(ectx);
}

static tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

	UNUSED(ectx);
	
    return tpm2_session_close(&ctx.key.object.session);
}


// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("verifydigestsignature", tpm2_tool_onstart, tpm2_tool_onrun, tpm2_tool_onstop, NULL)
